cpp/ql/lib/semmle/code/cpp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -0,0 +1,248 @@
+/**
+ * Provides classes and predicates for defining flow summaries.
+ */
+
+private import cpp
+private import codeql.dataflow.internal.FlowSummaryImpl
+private import codeql.dataflow.internal.AccessPathSyntax as AccessPath
+private import semmle.code.cpp.ir.dataflow.internal.DataFlowPrivate
+private import semmle.code.cpp.ir.dataflow.internal.DataFlowUtil
+private import semmle.code.cpp.ir.dataflow.internal.DataFlowImplSpecific as DataFlowImplSpecific
+private import semmle.code.cpp.dataflow.ExternalFlow
+
+module Input implements InputSig<DataFlowImplSpecific::CppDataFlow> {
+  class SummarizedCallableBase = Function;
+
+  ArgumentPosition callbackSelfParameterPosition() { result = TDirectPosition(-1) }
+
+  ReturnKind getStandardReturnValueKind() { result.(NormalReturnKind).getIndirectionIndex() = 0 }
+
+  string encodeParameterPosition(ParameterPosition pos) { result = pos.toString() }
+
+  string encodeArgumentPosition(ArgumentPosition pos) { result = pos.toString() }
+
+  string encodeReturn(ReturnKind rk, string arg) {
+    rk != getStandardReturnValueKind() and
+    result = "ReturnValue" and
+    arg = rk.toString()
+  }
```

Comment:
Since the `toString` on return kinds doesn't fully determine the return kind (because the indirection index isn't part of the `toString`) the MaD syntax will conflate all return kind indirections. That's probably fine for now, but we should fix this before claiming any MaD support 👍